tests/test_read_dtype.py

```diff
@@ -31,3 +72,55 @@ def test_float32_to_int16():
     assert data.dtype == numpy.dtype('int16')
     assert (data == numpy.array([[[ 1,  1,  0], [ 0,  0, -2]]], dtype='int16')).all()
 
+
+@pytest.mark.parametrize("dtype,nodata", itertools.product(DTYPES, [1, 127]))
+def test_write_mem(dtype, nodata):
+    profile = {
+        "driver": "GTiff",
+        "width": 2,
+        "height": 1,
+        "count": 1,
+        "dtype": dtype,
+        "crs": "EPSG:3857",
+        "transform": affine.Affine(10, 0, 0, 0, -10, 0),
+        "nodata": nodata,
+    }
+
+    values = numpy.array([[nodata, nodata]], dtype=dtype)
+
+    with rasterio.open("/vsimem/test.tif", "w", **profile) as src:
+        src.write(values, indexes=1)
+
+    with rasterio.open("/vsimem/test.tif") as src:
+        read = src.read(indexes=1)
+        assert read[0][0] == nodata
+        assert read[0][1] == nodata
+
+
+@pytest.mark.parametrize("dtype,nodata", itertools.product(DTYPES, [None, 1, 127]))
+def test_write_fs(tmp_path, dtype, nodata):
+    filename = tmp_path.joinpath("test.tif")
+    profile = {
+        "driver": "GTiff",
+        "width": 2,
+        "height": 1,
+        "count": 1,
+        "dtype": dtype,
+        "crs": "EPSG:3857",
+        "transform": affine.Affine(10, 0, 0, 0, -10, 0),
+        "nodata": nodata,
+    }
+
+    if dtype.startswith('int') or dtype.startswith('uint'):
+        info = numpy.iinfo(dtype)
+    else:
+        info = numpy.finfo(dtype)
+    values = numpy.array([[info.max, info.min]], dtype=dtype)
+
+    with rasterio.open(filename, "w", **profile) as src:
+        src.write(values, indexes=1)
+
+    with rasterio.open(filename) as src:
+        read = src.read(indexes=1)
+        assert read[0][0] == info.max
+        assert read[0][1] == info.min
```

Comment:
@adamjstewart I may be missing something. Why does the fs version of the test assert on info min and max while the mem version asserts on the nodata value?

Reply:
That's just how the existing tests in `tests/test_int8.py` were set up. Happy to change that if you want.

Reply:
@adamjstewart after reminding myself what the int8 issue was, I owe you an apology. Even Rouault says in https://github.com/OSGeo/gdal/issues/2298#issuecomment-595065049:

> Signed byte is not a dedicated data type in GDAL. There is some support in some places, but you shouldn't expect it to work everywhere properly. My advice is to gdal_translate to Int16 before doing any processing.

The int8 tests are a special case. I don't believe that I/O tests for all dtypes are needed. They don't test anything that isn't already tested by GDAL. I apologize for wasting your time, but I'm going to close this PR.

Reply:
> I don't believe that I/O tests for all dtypes are needed. They don't test anything that isn't already tested by GDAL.

If that was the case, I wouldn't have lost an entire weekend to #2502. #2467 introduced a very subtle bug that wasn't covered by a single unit test. Just because GDAL does the actual I/O doesn't mean rasterio is bug-free.